### PR TITLE
[SPARK-10914] UnsafeRow serialization breaks when two machines have different Oops size.

### PR DESCRIPTION
**Pull request by rxin**
UnsafeRow contains 3 pieces of information when pointing to some data in memory (an object, a base offset, and length). When the row is serialized with Java/Kryo serialization, the object layout in memory can change if two machines have different pointer width (Oops in JVM).

To reproduce, launch Spark using

MASTER=local-cluster[2,1,1024] bin/spark-shell --conf "spark.executor.extraJavaOptions=-XX:-UseCompressedOops"

And then run the following

scala> sql("select 1 xx").collect()

### CODE REVIEW

**Comment by rxin** [comment]
cc @davies and @JoshRosen

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoin.scala [hunk 1]
```diff
@@ -89,8 +89,13 @@ case class BroadcastHashJoin(
         // The following line doesn't run in a job so we cannot track the metric value. However, we
         // have already tracked it in the above lines. So here we can use
         // `SQLMetrics.nullLongMetric` to ignore it.
+
+        input.foreach { row =>
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoin.scala]: going to revert this

**Comment by davies** [comment]
LGTM, we have done this for UTF8String already (not support Kryo). 

@cloud-fan Should we also do this for UnsafeArrayData?

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java [hunk 2]
```diff
@@ -596,4 +601,40 @@ public boolean anyNull() {
...
+    this.baseObject = new byte[sizeInBytes];
+    in.readFully((byte[]) baseObject);
+  }
+
+  @Override
+  public void write(Kryo kryo, Output out) {
+    byte[] bytes = getBytes();
+    out.writeInt(bytes.length);
+    out.writeInt(this.numFields);
+    out.write(bytes);
```
Comment (andrewor14) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java]: can this just call `writeExternal(out)`?

File: sql/core/src/test/scala/org/apache/spark/sql/UnsafeRowSuite.scala [hunk 3]
```diff
@@ -29,6 +30,32 @@ import org.apache.spark.unsafe.types.UTF8String
...
+    // serializing an UnsafeRow pointing to a large buffer should only serialize the relevant data
+    val data = new Array[Byte](1024)
+    val row = new UnsafeRow
+    row.pointTo(data, 1, 16)
+    row.setLong(0, 19285)
+
+    val ser = new KryoSerializer(new SparkConf).newInstance()
+    val row1 = ser.deserialize[UnsafeRow](ser.serialize(row))
+    assert(row1.getLong(0) == 19285)
+    assert(row1.getBaseObject().asInstanceOf[Array[Byte]].length == 16)
```
Comment (andrewor14) [inline sql/core/src/test/scala/org/apache/spark/sql/UnsafeRowSuite.scala]: `===`

**Comment by cloud-fan** [comment]
I think we should apply this to unsafe array too.

File: sql/core/src/test/scala/org/apache/spark/sql/UnsafeRowSuite.scala (hunk 3 above)
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/UnsafeRowSuite.scala]: This actually doesn't matter anymore with new versions of ScalaTest.

**Comment by yhuai** [comment]
test this please

**Comment by rxin** [comment]
Merging this in master & branch-1.5.

**Pull request closed without merging**